arch/arm/core/aarch64/fatal.c

```diff
@@ -0,0 +1,176 @@
+		break;
+	case 0b011001:
+		LOG_ERR("Trapped access to SVE functionality");
+		break;
+	case 0b100000:
+		LOG_ERR("Instruction Abort from a lower Exception level, that "
+			"might be using AArch32 or AArch64");
+		break;
+	case 0b100001:
+		LOG_ERR("Instruction Abort taken without a change in Exception "
+			"level.");
+		break;
+	case 0b100010:
+		LOG_ERR("PC alignment fault exception.");
+		break;
+	case 0b100100:
+		LOG_ERR("Data Abort from a lower Exception level, that might "
+			"be using AArch32 or AArch64");
+		break;
+	case 0b100101:
+		LOG_ERR("Data Abort taken without a change in Exception level");
+		break;
+	case 0b100110:
+		LOG_ERR("SP alignment fault exception");
+		break;
+	case 0b101000:
+		LOG_ERR("Trapped floating-point exception taken from AArch32 "
+			"state");
+		break;
+	case 0b101100:
+		LOG_ERR("Trapped floating-point exception taken from AArch64 "
+			"state.");
+		break;
+	case 0b101111:
+		LOG_ERR("SError interrupt");
+		break;
+	case 0b110000:
+		LOG_ERR("Breakpoint exception from a lower Exception level, "
+			"that might be using AArch32 or AArch64");
+		break;
+	case 0b110001:
+		LOG_ERR("Breakpoint exception taken without a change in "
+			"Exception level");
+		break;
+	case 0b110010:
+		LOG_ERR("Software Step exception from a lower Exception level, "
+			"that might be using AArch32 or AArch64");
+		break;
+	case 0b110011:
+		LOG_ERR("Software Step exception taken without a change in "
+			"Exception level");
+		break;
+	case 0b110100:
+		LOG_ERR("Watchpoint exception from a lower Exception level, "
+			"that might be using AArch32 or AArch64");
+		break;
+	case 0b110101:
+		LOG_ERR("Watchpoint exception taken without a change in "
+			"Exception level.");
+		break;
+	case 0b111000:
+		LOG_ERR("BKPT instruction execution in AArch32 state");
+		break;
+	case 0b111100:
+		LOG_ERR("BRK instruction execution in AArch64 state.");
+		break;
+	}
+}
+
+void z_arm64_fatal_error(unsigned int reason)
+{
+	u64_t el, esr, elr, far;
+
+	if (reason != K_ERR_SPURIOUS_IRQ) {
+		__asm__ volatile("mrs %0, CurrentEL" : "=r" (el));
+
+		switch (GET_EL(el)) {
+		case MODE_EL1:
+			__asm__ volatile("mrs %0, esr_el1" : "=r" (esr));
+			__asm__ volatile("mrs %0, far_el1" : "=r" (far));
+			__asm__ volatile("mrs %0, elr_el1" : "=r" (elr));
+			break;
+		case MODE_EL2:
+			__asm__ volatile("mrs %0, esr_el2" : "=r" (esr));
+			__asm__ volatile("mrs %0, far_el2" : "=r" (far));
+			__asm__ volatile("mrs %0, elr_el2" : "=r" (elr));
+			break;
+		case MODE_EL3:
+			__asm__ volatile("mrs %0, esr_el3" : "=r" (esr));
+			__asm__ volatile("mrs %0, far_el3" : "=r" (far));
+			__asm__ volatile("mrs %0, elr_el3" : "=r" (elr));
+			break;
+		default:
+			/* Just to keep the compiler happy */
+			esr = elr = far = 0;
+			break;
+		}
+
+		if (GET_EL(el) != MODE_EL0) {
+			LOG_ERR("ESR_ELn: %llx", esr);
+			LOG_ERR("FAR_ELn: %llx", far);
+			LOG_ERR("ELR_ELn: %llx", elr);
+
+			print_EC_cause(esr);
+		}
+
+	}
+
+	z_fatal_error(reason, NULL);
```

Comment:
You should be dumping the register contents in the exception stack frame and passing that up to z_fatal_error. why is your esf NULL here?

Reply:
For a couple of reasons:
1. There is no exception stack frame for the aarch64 (that is something specific to Cortex-M). The registers in the `z_arch_esf_t` are only used when the thread is created to carry the thread entry point and the parameters. Most of the useful information in case of error is carried anyway in the `ESR`, `FAR` e `ELR` registers that we print as part of the error routine.
2. We pass `NULL` to `z_fatal_error()` because the parameter is not actually used.

Reply:
> There is no exception stack frame for the aarch64 (that is something specific to Cortex-M). The registers in the z_arch_esf_t are only used when the thread is created to carry the thread entry point and the parameters. Most of the useful information in case of error is carried anyway in the ESR, FAR e ELR registers that we print as part of the error routine.

So to make sure I understand you correctly, when an aarch64 CPU takes an exception, nothing gets pushed onto the stack, and the exception handling code doesn't save any state either? Is this because the arch does some kind of register bank switching scheme or similar magic behind the scenes, or is this interrupted thread state simply discarded under the assumption that exceptions are non-recoverable?

> Most of the useful information in case of error is carried anyway in the ESR, FAR e ELR registers that we print as part of the error routine.

These seem to cover the PC value and exception type, is there any way to dump the stack pointer when the exception occurred?

> We pass NULL to z_fatal_error() because the parameter is not actually used.

Strictly speaking it gets passed along to k_sys_fatal_error_handler(), which has a trivial default implementation but in a production system might have logic that is interested to know the context in which a fatal error occurred. You might consider packaging up the values that can be known (ESR, FAR, ELR, etc) as the esf, despite its name, is intended to be an architecture-specific structure containing information about the exception when it happened. Could be a future enhancement, not necessarily in this PR.

Reply:
> So to make sure I understand you correctly, when an aarch64 CPU takes an exception, nothing gets pushed onto the stack, and the exception handling code doesn't save any state either? 

Nothing is automatically pushed onto the stack. What happens is only that `PSTATE` is saved into `SPSL_ELx` and `PC` is saved into `ELR_ELx`, nothing else.

The exception handling code in this port saves these two registers and all the other needed registers on the thread stack in case the thread is switched out.

> Is this because the arch does some kind of register bank switching scheme or similar magic behind the scenes, or is this interrupted thread state simply discarded under the assumption that exceptions are non-recoverable?

There is no switching or magic, everything is left to the implementer. If you want to save the registers onto the stack when entering the exception to be recovered later is up to you.

A bit of help comes from the fact that each exception has its own dedicated stack pointer, so you can use a different dedicated `SP` for the exception (so on entering the exception you can save the registers corrupted by the exception handler on the thread `SP` and retrieve those in the error handler while keep using a different `SP` for the exception). This is totally left to the programmer though and it is not currently implemented. In the current implementation for the sake of simplicity we are using one single `SP` (`SP_ELn`) for threads and exceptions.

I can check what's needed to have it implemented but I prefer to have this done as a separate PR.

> These seem to cover the PC value and exception type, is there any way to dump the stack pointer when the exception occurred?

As detailed before we could switch to a different `SP` when taking the exception leaving the thread `SP` intact but this requires quite a bit of changes in the core code that I prefer to implement at a later stage.

Reply:
> As detailed before we could switch to a different `SP` when taking the exception leaving the thread `SP` intact 

Not this way but yes the purpose is same. I am just clarifying one behavior of core for others.

Exception handlers are always entered in handler mode. ie Core switches to SP_ELn on exception automatically.  (regardless of SPSEL settings).  Something similar to MSP in cortex-M.  One have choice to manually switch to SP_EL0 (thread mode) (like PSP in cortex-M) to run rest of the code Or to continue with SP_ELn.  Ex: Linux continues to use SP_ELn.  So this feature helps having a separate reserved more reliable stack for exception handlers immune from rest of the code.  Linux is still not using this scheme though.

I agree this is a bit non trivial change to use thread mode and really can be deferred to be decided later.

Reply:
Yes, thank you for clarifying this. In the reset code we switch to `SP_ELn` and we stick to that.

Reply:
Looking again at `swap_helper.S` when we end up into the error handling routing the thread stack is untouched, so even thought technically that is not the ESF we can still recover the registers content at the time of the exception. I'll try to cook a patch for that (without touching the core code).

Reply:
> A bit of help comes from the fact that each exception has its own dedicated stack pointer, so you can use a different dedicated SP for the exception (so on entering the exception you can save the registers corrupted by the exception handler on the thread SP and retrieve those in the error handler while keep using a different SP for the exception). This is totally left to the programmer though and it is not currently implemented. In the current implementation for the sake of simplicity we are using one single SP (SP_ELn) for threads and exceptions.

I'd prefer to have a `TODO` in place where we are expecting some obvious improvements like this one. This will help other developers who wish to improve the port in future.

Reply:
> A bit of help comes from the fact that each exception has its own dedicated stack pointer, so you can use a different dedicated SP for the exception (so on entering the exception you can save the registers corrupted by the exception handler on the thread SP and retrieve those in the error handler while keep using a different SP for the exception). This is totally left to the programmer though and it is not currently implemented. In the current implementation for the sake of simplicity we are using one single SP (SP_ELn) for threads and exceptions.

I'd prefer to have a TODO in place where we are expecting some obvious improvements like this one. This will help other developers who wish to improve the port in future.